REP-5758 JSON to YAML CLI Tool

dmnjohns opened the pull request:
I believe this satisfies the acceptance criteria for the story, but it does not do much bounds checking.

RackerWilliams commented on core/src/main/scala/com/rackspace/identity/components/AttributeMapper.scala:
```diff
@@ -366,6 +366,23 @@ object AttributeMapper {
   }
 
+  // Translates a JSON policy to a YAML policy.
+  def policy2YAML(policy : InputStream, policyYAML : OutputStream, validate : Boolean, xsdEngine : String) : Unit = {
```
For consistency's  sake you should be using StreamSource here.  You already have a call that parses the YAML and turns it into JSON -- it's called parseYAMLNode  -- you should be calling that instead of repeating the logic here.

dmnjohns replied:
Fair enough. I'll update that.

In that case, would you still want to take an `OutputStream` as an argument to write to, or would just returning a `String` or the raw UTF-8 encoded `Byte`s be better? I stuck with an `OutputStream` to maintain some semblance of consistency with the other conversion methods, but I don't know of a way to write to a `Destination` since we aren't doing any XSLT processing.

dmnjohns replied:
As a side note, I lean toward returning either a `String` or `Array[Byte]` as that is more functional, but then the behavior isn't as consistent with the other conversion methods.

RackerWilliams replied:
I'd leave it as output stream.  You close a couple of doors if you go byteArray or string.

RackerWilliams commented on core/src/main/scala/com/rackspace/identity/components/AttributeMapper.scala:
```diff
@@ -366,6 +366,23 @@ object AttributeMapper {
 
+  // Translates a JSON policy to a YAML policy.
+  def policy2YAML(policy : InputStream, policyYAML : OutputStream, validate : Boolean, xsdEngine : String) : Unit = {
+    val jsonMapper = new ObjectMapper()
```
...additionally on line 267 instead of saying

````scala
 policy2XML(parseYamlNode(policy.asInstanceOf[StreamSource]), outPolicyXML)
````

You should be saying

````scala
 policy2XML(policy2YAML(policy.asInstanceOf[StreamSource],...,), outPolicyXML)
````

You don't have to repeat yourself..plus your code will now be exercised in all of the tests.

dmnjohns replied:
Huh? On line 267, we already have YAML coming off the stream. `policy2XML` only supports YAML if passed a `JsonNode`, so calling `parseYamlNode` will parse the YAML and return a `JsonNode` for us to use. Plus, `policy2YAML` writes to a stream so that it behaves similarly to the other conversion methods which write to some `Destination`, so you would have to use some intermediary state to get the chaining to work.

What I kind of wanted to do, and hinted at in the TODO file, is expand the conversion methods to be able to take any supported format (i.e., YAML, JSON, XML) and return the desired format. That felt like it was outside the scope of this story, though.

RackerWilliams replied:
Ugh, you're right.  That's what I get for doing a quick review between meetings.  I'm confused.  Disregard my comment.

RackerWilliams commented on core/src/test/scala/com/rackspace/identity/components/AttributeMapperTest.scala:
```diff
@@ -0,0 +1,143 @@
+/**
```
While nice, I don't thing these tests are necessary if you make the changes above.  Why?

1.  There is a JSON policy in the test suite
2.  We expect the YAML version of the policy to behave exactly as the JSON version does, so we should be good conversion wise.

If you do want to go with these tests, I'm afraid they may be brittle because you're doing direct string comparisons -- instead of comparing JSON nodes.   Does ```stripMargin``` strip all unneeded spaces? If not, then a change in Jackson that adds an extra space may break these tests.

dmnjohns replied:
So there is a JSON policy in the test suite, but the test suite does not test the JSON to YAML conversion. It only tests the YAML to JSON conversion as part of testing policies in general.

Also, these tests are not doing direct string comparisons -- it is comparing the object generated from parsing the output to the object generated from parsing the expected output.

RackerWilliams replied:
Again, I was confused, and obviously didn't review to closely.  Disregard.

RackerWilliams commented on cli/attribmap2yaml/src/main/scala/com/rackspace/identity/components/cli/AttribMap2YAML.scala:
```diff
@@ -0,0 +1,117 @@
+                                             "Display version.")
+
+
+    def policyStream : InputStream = new FileInputStream(URLResolver.toAbsoluteSystemId(policy.value.get, base))
```
This isn't working.  You're not resolving local files correctly.  Not sure where the error is, but I bet it has something to do with setting the base?

Here's what I see when jumping into the directory:  ```/Users/jorgew/projects/attributeMapping/core/src/test/resources/tests/mapping-tests/name-and-roles-xpath/maps``` and running ```attribmap2xml``` and ```attribmap2yaml```.  Notice the XML conversion works but the json one fails with an error.


````shell
jorge-williamss-macbook-pro:maps jorgew$ attribmap2xml  name-and-roles-xpath-s.json
<?xml version="1.0" encoding="UTF-8"?>
<mapping xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xmlns:xs="http://www.w3.org/2001/XMLSchema"
         xmlns="http://docs.rackspace.com/identity/api/ext/MappingRules"
         version="RAX-1">
   <rules>
      <rule>
         <local>
            <user>
               <domain value="78334"/>
               <name value="{D}"/>
               <email value="{D}"/>
               <roles value="{Pts(if ((xs:integer(seconds-from-time(current-time())) mod 2) = 0) then ('nova:admin', 'lbaas:admin') else ('nova:observer','lbaas:observer'))}"
                      multiValue="true"/>
               <expire value="{D}"/>
            </user>
         </local>
      </rule>
   </rules>
</mapping>jorge-williamss-macbook-pro:maps jorgew$ attribmap2yaml  name-and-roles-xpath-s.yaml
java.io.FileNotFoundException: /Users/jorgew/projects/attributeMapping/core/src/test/resources/tests/mapping-tests/name-and-roles-xpath/name-and-roles-xpath-s.yaml (No such file or directory)
	at java.io.FileInputStream.open0(Native Method)
	at java.io.FileInputStream.open(FileInputStream.java:195)
	at java.io.FileInputStream.<init>(FileInputStream.java:138)
	at java.io.FileInputStream.<init>(FileInputStream.java:93)
	at com.rackspace.identity.components.cli.AttribMap2YAML$.policyStream$1(AttribMap2YAML.scala:59)
	at com.rackspace.identity.components.cli.AttribMap2YAML$.parseArgs(AttribMap2YAML.scala:79)
	at com.rackspace.identity.components.cli.AttribMap2YAML$.nailMain(AttribMap2YAML.scala:109)
	at com.rackspace.identity.components.cli.AttribMap2YAML.nailMain(AttribMap2YAML.scala)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.martiansoftware.nailgun.NGSession.run(NGSession.java:280)
````

dmnjohns replied:
Ah. So yeah, you're right. I had some issues with the base setting logic, and removing it works... as long as you're in the same directory as the script. I'll take another look and get this fixed.

dmnjohns replied:
So everything worked once I switched over to using a `StreamSource`. When I was trying to use a `FileInputStream`, the `FileInputStream` complained that the file did not exist. The `String` being passed to the `FileInputStream` was something like `file:/Users/<me>/dev/attributeMapping/bin`. The `FileInputStream` was perfectly happy to accept `/Users/<me>/dev/attributeMapping/bin`, though, so something about the protocol threw it off whereas the `StreamSource` had no issues.

RackerWilliams replied:
Yea, the XML APIs suck.  But one of the nice things about them is that they get a couple of things right in terms of putting together URIs with relevant paths.

dmnjohns commented:
Huh. So I'm guessing that's the result of attempting to convert the post-validation content. Let me play around with that and figure out exactly what's going on.

RackerWilliams commented:
You are right!  This is a bug in converting XML to JSON.   Let me dig into that one.  I may have to patch that before I accept this request.

RackerWilliams commented:
When I added support to Pt({...}),  and Pts({...}).  I consider the conversions from JSON->XML, since that's the one that'll activate at runtime.  But I didn't take into account the conversion from XML->JSON.  That's the issue.  I'm on it.

dmnjohns commented:
Alright! In that case, I'll leave that to you! Let me know if I have any work to do for it.

I am going to add another test case for conversion and validation, since I apparently missed that one.

RackerWilliams commented:
Okay I have a fix in this branch:

https://github.com/RackerWilliams/attributeMapping/tree/mapping_to_json_bug

I need to write tests around it, and I won't get to do that till tonight. As soon as I have good tests, I'll send my fix for a PR, once that goes through rebase and I'll review.

I should have example conversion tests that you can follow for YAML in my PR.

RackerWilliams commented:
Okay, fixed in #37.  Please review.  Example of conversion tests suite is there as well.

dmnjohns commented:
I have rebased onto the changes from #37 and added a JSON to YAML conversion test suite. With that suite in place, do you want to keep the `AttributeMapperTest`? That is more of a unit test, and includes a test expecting an exception if validation is enabled, but the input is invalid. Other than that, the conversion suite should suffice, and test the same cases.

RackerWilliams commented:
I'm okay keeping ```AttributeMapperTest```

dmnjohns commented:
Alright. I made a small mistake in the tests, but that should be fixed now. As soon as the tests pass, I think this is ready for re-review.

RackerWilliams commented on core/src/test/resources/tests/mapping-tests/adfs-faws-ext2/maps/faws.json:
```diff
@@ -39,14 +39,14 @@
                 "AdminstratorAccess"
                ],
-              "multiValue":true
+               "multiValue": true
```
Why are we modifying this?

dmnjohns replied:
Because I'm bad formatting and I should feel bad. I'll remove that change.
